TestScatter: span used after call to Finish when running at high verbosity

After applying #15355 to fix #15003, I tried stressing `TestScatter` hard to verify it fixed the problem. I happened to be running with `-verbosity=3`, and ran into this panic:

```
E170426 15:43:13.477420 2711 util/log/stderr_redirect.go:66  [n1] a panic has occurred!
panic: span used after call to Finish():
&{tracer:<nil> event:0xccdd80 Mutex:{state:0 sema:0} raw:{Context:{TraceID:8535667686755371615 SpanID:2911314615026932369 Sampled:false Baggage:map[]} ParentSpanID:9096957308716108751 Operation:node.Batch Start:{sec:63628818192 nsec:306743425 loc:0x3315380} Duration:1137823811 Tags:map[] Logs:[]} numDroppedLogs:0} [recovered]
	panic: span used after call to Finish():
&{tracer:<nil> event:0xccdd80 Mutex:{state:0 sema:0} raw:{Context:{TraceID:8535667686755371615 SpanID:2911314615026932369 Sampled:false Baggage:map[]} ParentSpanID:9096957308716108751 Operation:node.Batch Start:{sec:63628818192 nsec:306743425 loc:0x3315380} Duration:1137823811 Tags:map[] Logs:[]} numDroppedLogs:0} [recovered]
	panic: span used after call to Finish():
&{tracer:<nil> event:0xccdd80 Mutex:{state:0 sema:0} raw:{Context:{TraceID:8535667686755371615 SpanID:2911314615026932369 Sampled:false Baggage:map[]} ParentSpanID:9096957308716108751 Operation:node.Batch Start:{sec:63628818192 nsec:306743425 loc:0x3315380} Duration:1137823811 Tags:map[] Logs:[]} numDroppedLogs:0} [recovered]
	panic: span used after call to Finish():
&{tracer:<nil> event:0xccdd80 Mutex:{state:0 sema:0} raw:{Context:{TraceID:8535667686755371615 SpanID:2911314615026932369 Sampled:false Baggage:map[]} ParentSpanID:9096957308716108751 Operation:node.Batch Start:{sec:63628818192 nsec:306743425 loc:0x3315380} Duration:1137823811 Tags:map[] Logs:[]} numDroppedLogs:0}

goroutine 2711 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4201806e0, 0x7f01f10bc600, 0xc422d53080)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:200 +0x104
panic(0x1ef4880, 0xc42357b240)
	/home/alex/go1.8/src/runtime/panic.go:489 +0x2f0
github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send.func1(0xc422e15420, 0xc422e15498, 0x14b8fcf1d74ac810, 0x0, 0xc422e15490, 0xc4208fc400)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2425 +0x57b
panic(0x1ef4880, 0xc42357b240)
	/home/alex/go1.8/src/runtime/panic.go:489 +0x2f0
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges.func1(0xc422e13798, 0xc422e13de0, 0xc422e13dd8)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:684 +0x3d3
panic(0x1ef4880, 0xc42357b240)
	/home/alex/go1.8/src/runtime/panic.go:489 +0x2f0
github.com/cockroachdb/cockroach/vendor/github.com/opentracing/basictracer-go.(*spanImpl).maybeAssertSanityLocked(0xc421faa3c0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/opentracing/basictracer-go/debug.go:31 +0x443
github.com/cockroachdb/cockroach/vendor/github.com/opentracing/basictracer-go.(*spanImpl).Lock(0xc421faa3c0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/opentracing/basictracer-go/debug.go:25 +0x4d
github.com/cockroachdb/cockroach/vendor/github.com/opentracing/basictracer-go.(*spanImpl).LogFields(0xc421faa3c0, 0xc42341dec0, 0x1, 0x1)
	/home/alex/go/src/github.com/cockroachdb/cockroach/vendor/github.com/opentracing/basictracer-go/span.go:123 +0xf3
github.com/cockroachdb/cockroach/pkg/util/log.eventInternal(0x7f01f10bc600, 0xc421c3a0f0, 0x1e90000, 0x20c858e, 0x8, 0xc422e12a88, 0x3, 0x3)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:132 +0x2cb
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x7f01f10bc600, 0xc421c3a0f0, 0x1, 0x2, 0x210358d, 0x2a, 0xc422e12c90, 0x3, 0x3)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:145 +0x2c6
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x7f01f10bc600, 0xc421c3a0f0, 0x1, 0x1, 0x210358d, 0x2a, 0xc422e12c90, 0x3, 0x3)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:68 +0x9a
github.com/cockroachdb/cockroach/pkg/util/log.Infof(0x7f01f10bc600, 0xc421c3a0f0, 0x210358d, 0x2a, 0xc422e12c90, 0x3, 0x3)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:86 +0x90
github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).evictCachedRangeDescriptorLocked(0xc420082be0, 0x7f01f10bc600, 0xc421c3a0f0, 0xc4237bae93, 0x0, 0xd, 0xc423468d80, 0xc422e12d00, 0x572d22, 0xc421b09580)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:452 +0x2fe
github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3.1(0xc42341d4e8, 0x2ecf1a8)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:300 +0xe2
github.com/cockroachdb/cockroach/pkg/kv.(*EvictionToken).EvictAndReplace.func1()
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:210 +0xfc
sync.(*Once).Do(0xc42341d4c8, 0xc422e12e80)
	/home/alex/go1.8/src/sync/once.go:44 +0xe2
github.com/cockroachdb/cockroach/pkg/kv.(*EvictionToken).EvictAndReplace(0xc42341d4c0, 0x7f01f10bc600, 0xc42225c780, 0x0, 0x0, 0x0, 0x41, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:219 +0x127
github.com/cockroachdb/cockroach/pkg/kv.(*EvictionToken).Evict(0xc42341d4c0, 0x7f01f10bc600, 0xc42225c780, 0x41, 0xc422e13140)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:196 +0x6c
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch(0xc42047c700, 0x7f01f10bc600, 0xc42225c780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:965 +0x774
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges(0xc42047c700, 0x7f01f10bc600, 0xc42225c780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:814 +0xb6b
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send(0xc42047c700, 0x7f01f10bc600, 0xc42225c780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:629 +0x3ca
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send(0xc4201f0f20, 0x7f01f10bc600, 0xc42225c6c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:456 +0x490
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send(0xc42000c640, 0x7f01f10bc600, 0xc42354d860, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:528 +0x228
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).(github.com/cockroachdb/cockroach/pkg/internal/client.send)-fm(0x7f01f10bc600, 0xc42354d860, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:464 +0x97
github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill(0x7f01f10bc600, 0xc42354d860, 0xc422e14830, 0xc420ef5500, 0x28, 0x1fd4340)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:436 +0x13b
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run(0xc42000c640, 0x7f01f10bc600, 0xc42354d860, 0xc420ef5500, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:464 +0xf8
github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).resolveIntents.func1(0xc42225c6b0, 0xc42225c690)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:431 +0xb3
github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).resolveIntents(0xc42055aba0, 0x7f01f10bc600, 0xc42354d860, 0xc422a47c20, 0x1, 0x1, 0x14b8fcf1d74a0000, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:442 +0x851
github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).processWriteIntentError(0xc42055aba0, 0x7f01f10bc600, 0xc42354d860, 0xc4216b33b0, 0x2ee38c0, 0xc4233b3230, 0x14b8fcf1d74ac810, 0x0, 0x100000001, 0x1, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:93 +0x1c0
github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send(0xc4208fc400, 0x7f01f10bc600, 0xc42354d860, 0x14b8fcf1d6d8baf0, 0x0, 0x100000001, 0x1, 0x1, 0x0, 0xc4229540d0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2602 +0xed6
github.com/cockroachdb/cockroach/pkg/storage.(*Stores).Send(0xc42056e100, 0x7f01f10bc600, 0xc42354d740, 0x0, 0x0, 0x100000001, 0x1, 0x1, 0x0, 0xc4229540d0, ...)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/storage/stores.go:187 +0x24b
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0x7f01f10bc600, 0xc42354d740, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:840 +0x20f
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc4201806e0, 0x7f01f10bc600, 0xc422d53080, 0xc422e157d8, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:272 +0x14f
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc4201a8580, 0x7f01f10bc600, 0xc422d53080, 0xc422df4028, 0xc422d53080, 0x1, 0xc400000001)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:851 +0x1c1
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc4201a8580, 0x7f01f10bc600, 0xc422d53080, 0xc422df4028, 0x1f, 0x0, 0x0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:868 +0xb8
github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext.func1.1(0xc42355b540, 0xc422df4000, 0x7f01f10bc600, 0xc4233b3890, 0xc4210c1f80, 0xc58ef1, 0xc4210c1f90)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:252 +0x73d
github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext.func1(0xc42355b540, 0xc422df4000, 0x7f01f10bc600, 0xc4233b3890, 0x2ee6d80, 0xc422c235e0, 0xc4230ee3c0)
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:265 +0xdd
created by github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext
	/home/alex/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:278 +0x236
```

It looks like it'll only trigger at v=2 or above, so it isn't a huge deal, but I'll send out a fix.

@andreimatei

--- a/pkg/kv/range_cache.go
+++ b/pkg/kv/range_cache.go
@@ -175,12 +175,12 @@ type EvictionToken struct {
 
 	doOnce    sync.Once                                               // assures that do and doReplace are run up to once.
 	doLocker  sync.Locker                                             // protects do and doReplace.
-	do        func() error                                            // called on eviction.
+	do        func(context.Context) error                             // called on eviction.
 	doReplace func(context.Context, ...roachpb.RangeDescriptor) error // called after eviction on EvictAndReplace.
 }
 
 func (rdc *RangeDescriptorCache) makeEvictionToken(
-	prevDesc *roachpb.RangeDescriptor, evict func() error,
+	prevDesc *roachpb.RangeDescriptor, evict func(ctx context.Context) error,
 ) *EvictionToken {
 	return &EvictionToken{
 		prevDesc:  prevDesc,
@@ -207,7 +207,7 @@ func (et *EvictionToken) EvictAndReplace(
 	et.doOnce.Do(func() {
 		et.doLocker.Lock()
 		defer et.doLocker.Unlock()
-		err = et.do()
+		err = et.do(ctx)
 		if err == nil {
 			if len(newDescs) > 0 {
 				err = et.doReplace(ctx, newDescs...)
@@ -269,7 +269,7 @@ func (rdc *RangeDescriptorCache) lookupRangeDescriptorInternal(
 		return nil, nil, err
 	} else if desc != nil {
 		rdc.rangeCache.RUnlock()
-		returnToken := rdc.makeEvictionToken(desc, func() error {
+		returnToken := rdc.makeEvictionToken(desc, func(ctx context.Context) error {
 			return rdc.evictCachedRangeDescriptorLocked(ctx, key, desc, useReverseScan)
 		})
 		return desc, returnToken, nil
@@ -296,7 +296,7 @@ func (rdc *RangeDescriptorCache) lookupRangeDescriptorInternal(
 			desc := &rs[0]
 			lookupRes = lookupResult{
 				desc: desc,
-				evictToken: rdc.makeEvictionToken(desc, func() error {
+				evictToken: rdc.makeEvictionToken(desc, func(ctx context.Context) error {
 					return rdc.evictCachedRangeDescriptorLocked(ctx, key, desc, useReverseScan)
 				}),
 			}
@@ -305,7 +305,7 @@ func (rdc *RangeDescriptorCache) lookupRangeDescriptorInternal(
 			nextDesc := rs[1]
 			lookupRes = lookupResult{
 				desc: desc,
-				evictToken: rdc.makeEvictionToken(desc, func() error {
+				evictToken: rdc.makeEvictionToken(desc, func(ctx context.Context) error {
 					return rdc.insertRangeDescriptorsLocked(ctx, nextDesc)
 				}),
 			}

